### PR TITLE
Track method's visibility during indexing

### LINKED ISSUE #1941: Save method visibilities in the index
The main cases to support are:

```rb
class Foo
  private # or protected

  def bar; end

  private def baz; end # or protected
end
```

### PR DESCRIPTION
**Pull request by st0012**
### Motivation

Closes #1941

### Implementation

- Adding a visibility field to the Member entry, which is a superclass of Accessor, Method, and Alias.
- In the DeclarationListener, create a nested stack structure to track the visibility of the current scope.

#### Nested stack

Every namespace will have its own visibility stack, for example:

```rb
class Foo
  class Bar
  end
end

# [[<top-level visibility stack>], [<Foo visibility stack>], [<Foo::Bar visibility stack>]]
```

And at each level (scope), we simply push new visibility to its stack whenever we see a modifier.

```rb

 # [[:public]]

def foo; end # :public

private # [[:public, :private]]

def bar; end # :private

protected # [[:public, :private, :protected]]

def baz; end :protected
```

The only exception is when we face `private def foo; end`, which will pop the stack when we leave the declaration.

I chose a different data structure than [test `CodeLens`'s](https://github.com/Shopify/ruby-lsp/blob/main/lib/ruby_lsp/listeners/code_lens.rb#L41) because it assumes methods are always inside a class, which would create a problem here. And I also feel the way it [swaps visibilities](https://github.com/Shopify/ruby-lsp/blob/65752c45ac5f0539e21fee3c8509235e96898982/lib/ruby_lsp/listeners/code_lens.rb#L144-L145) is harder to understand than the nested stack I use here.

One drawback of this design is that if one scope level has many visibility modifiers, it'd create a big visibility stack at that level, like:

```rb
private
private
private
# repeat 997 more times, then we'd have [[:private * 1000]]
```

However, I think it's an edge case that we can ignore in order to get a simpler implementation.

### Automated Tests

I added 2 new test cases for this change.

### CODE REVIEW

**Comment by st0012** [comment]
@vinistock I've changed the way visibility is set on entries and reduced the amount of changes needed.

File: lib/ruby_indexer/lib/ruby_indexer/declaration_listener.rb [hunk 1]
```diff
@@ -50,6 +52,7 @@ def initialize(index, dispatcher, parse_result, file_path)
 
     sig { params(node: Prism::ClassNode).void }
     def on_class_node_enter(node)
```
[the remark this answers is not in the document]
Comment (vinistock) [reply lib/ruby_indexer/lib/ruby_indexer/declaration_listener.rb:54]: We are not handling singleton contexts properly anywhere. We'll need to take a detailed look into the entire indexing approach for singletons (which is likely one of the next immediate tasks).

File: lib/ruby_indexer/lib/ruby_indexer/declaration_listener.rb [hunk 2]
```diff
@@ -11,6 +11,7 @@ class DeclarationListener
     def initialize(index, dispatcher, parse_result, file_path)
       @index = index
       @file_path = file_path
+      @visibility_stack = T.let([[:public]], T::Array[T::Array[Symbol]])
```
[the remark this answers is not in the document]
Comment (vinistock) [reply lib/ruby_indexer/lib/ruby_indexer/declaration_listener.rb:14]: It's for nested classes and modules. You need to remember the stack of visibilities inside each level of nesting in case you have something like this

```ruby
module Foo
  private
  def bar; end

  module Bar
    def baz; end # public!
  end

  def qux; end # private again
end
```

**Review by vinistock — APPROVED** [review]
I agree with Alex. I would reduce the indirection in favour of more direct code.

File: lib/ruby_indexer/lib/ruby_indexer/declaration_listener.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (Morriar) [reply lib/ruby_indexer/lib/ruby_indexer/declaration_listener.rb:14]: Yeah I understand why the stack, what I don't understand is the stack of arrays: `T::Array[T::Array[Symbol]])`

[the remark this answers is not in the document]
Comment (vinistock) [reply lib/ruby_indexer/lib/ruby_indexer/declaration_listener.rb:14]: Oh, I see what you mean. Yeah, I guess it could just be a flat array indeed.

[the remark this answers is not in the document]
Comment (Morriar) [reply lib/ruby_indexer/lib/ruby_indexer/declaration_listener.rb]: Why can't you just push the inline visibility in the stack?

**Review by Morriar — APPROVED** [review]
This is great! 🎉
